jobrunner/job_executor.py

```diff
@@ -33,26 +33,100 @@ class JobDefinition:
     allow_database_access: bool  # whether this job should have access to the database
 
 
+class ExecutorState(Enum):
+    PREPARING = "preparing"
+    PREPARED = "prepared"
+    EXECUTING = "executing"
+    EXECUTED = "executed"
+    FINALIZING = "finalizing"
+    FINALIZED = "finalized"
+    UNKNOWN = "unknown"
+    ERROR = "error"
+
+
+@dataclass
+class JobStatus:
+    state: ExecutorState
+    message: Optional[str] = None
```

Comment:
A thing it's worth bearing in mind is some mechanism for signalling specific error states that doesn't rely on parsing a free-form `message` field. In the old system we had the (not very well-named) `status_code` field for that. We could just do this by convention here e.g. format message as `[<error code>] <string>`. No strong opinions (yet!) on the right approach, but I think we'll need something like to build a properly helpful UI.

Reply:
So, I went without codes, as the specific issues that the codes denote were able to determined by state logic alone, w/o a code.

And we only currently *need* the code to communicate w/job-server, and I was keen to keep job-server specific things out of the executor, so that we can change them easily.

That said, there may be future issues where the executor returns a state that could imply different things that need a code to determine, but right now there aren't any. If that occurs we can model it either as a new state, or add a code, as which ever is appropriate.  We shouldn't never end up parsing a message. We could add a `code` to the `JobStatus` now, with `# reserved for future use`?

Reply:
Ah, you're quite right that all the current ones are things you can infer from other state. I keep confusing "things I was planning to do" with "things I have actually done" :smile: The _intention_ was that we'd have a way of signalling things like "this was on out-of-memory error", "this was an our of disk-space error" etc to job-server so it could display a more helpful error message, perhaps with links to advice on what to do next.

Maybe the best thing to do for now is to keep the API simple and worry about how to extend it when we have a concrete need. It's not like we're setting it in stone forever at this point.